pkg/services/ngalert/api/api_alertmanager.go

```diff
@@ -343,8 +350,99 @@ func (srv AlertmanagerSrv) RoutePostTestReceivers(c *contextmodel.ReqContext, bo
 }
 
 func (srv AlertmanagerSrv) RoutePostTestTemplates(c *contextmodel.ReqContext, body apimodels.TestTemplatesConfigBodyParams) response.Response {
-	// TODO: Stub
-	return response.JSON(http.StatusOK, apimodels.TestTemplatesResults{})
+	am, errResp := srv.AlertmanagerFor(c.OrgID)
+	if errResp != nil {
+		return errResp
+	}
+
+	paths := make([]string, 0)
+	for _, name := range am.GetStatus().Config.Templates {
+		if name == body.Name {
+			// Skip the existing template of the same name as we're going to parse the one for testing instead.
+			continue
+		}
+		paths = append(paths, filepath.Join(am.WorkingDirPath(), name))
+	}
+
+	res, err := TestTemplate(c.Req.Context(), am.AppURL(), body, paths, srv.log)
+	if err != nil {
+		return response.Error(http.StatusInternalServerError, "", err)
+	}
+
+	return response.JSON(http.StatusOK, res)
+}
+
+func TestTemplate(ctx context.Context, appUrl string, c apimodels.TestTemplatesConfigBodyParams, paths []string, l log.Logger) (*apimodels.TestTemplatesResults, error) {
+
+	definitions := make([]string, 0)
+	var extractDefinitions template.Option = func(text *tmpltext.Template, html *tmplhtml.Template) {
+		for _, t := range text.Templates() {
+			if t.Tree == nil || t.Root == nil {
+				continue
+			}
+			definitions = append(definitions, t.Name())
```

Comment:
Don't forget we need to exclude nested templates that are called from other templates 🙂